### PR TITLE
[query] SeqSample

### PR DESCRIPTION
**Pull request by johnc1231**
This PR adds `SeqSample(totalRange, numToSample)`, a stream IR that generates a stream of `numToSample` unique random integers in the range `totalRange`.

### CODE REVIEW

File: hail/src/main/scala/is/hail/expr/ir/streams/EmitStream.scala [hunk 1]
```diff
@@ -452,6 +452,55 @@ object EmitStream {
...
+
+            val nRemaining = cb.newLocal[Int]("seq_sample_num_remaining", numToSampleVal.intCode(cb))
+            val candidate = cb.newLocal[Int]("seq_sample_candidate", 0)
+            val elementToReturn = cb.newLocal[Int]("seq_sample_element_to_return", -1) // -1 should never be returned.
+
+            val producer = new StreamProducer {
+              override val length: Option[EmitCodeBuilder => Code[Int]] = Some(_ => len)
+
+              override def initialize(cb: EmitCodeBuilder): Unit = {
+                cb.assign(len, numToSampleVal.asInt.intCode(cb))
```
Comment (patrick-schultz) [inline hail/src/main/scala/is/hail/expr/ir/streams/EmitStream.scala:474]: I think you need to initialize all your state variables here (nRemaining, candidate, elementToReturn)

File: hail/src/test/scala/is/hail/expr/ir/EmitStreamSuite.scala [hunk 2]
```diff
@@ -193,6 +193,19 @@ class EmitStreamSuite extends HailSuite {
...
+    val seq = evalStream(
+      SeqSample(
+        I32(20),
+        I32(5),
+        false
+      )
+    ).map(_.asInstanceOf[Int])
+    assert(seq.size == 5)
+    assert(seq.toSet.size == 5)
+    assert(seq.forall(e => e >= 0 && e < 20))
```
Comment (patrick-schultz) [inline hail/src/test/scala/is/hail/expr/ir/EmitStreamSuite.scala:206]: It's always hard to test random methods, but we should probably have something to check that this has the right statistical properties. I would suggest repeating `SeqSample(N, n)` a large number of times (maybe compile once and run repeatedly), where n is 2 or 3, and N is smallish, then count how many times each possible two element subset gets sampled in an NxN matrix. If the sampler is uniform, they should all be nearly equal.

**Review by patrick-schultz — CHANGES_REQUESTED** [review]
Thanks John! Just a couple things.

**Comment by johnc1231** [comment]
Double check that the test is what you want. I made the bounds pretty wide to make sure we never had an unlucky sporadic test failure.

File: hail/src/test/scala/is/hail/expr/ir/EmitStreamSuite.scala [hunk 3]
```diff
@@ -193,6 +193,38 @@ class EmitStreamSuite extends HailSuite {
...
+      assert(seq.toSet.size == n)
+      assert(seq.forall(e => e >= 0 && e < N))
+    }
+
+
+    (0 until N).foreach { i =>
+      (i + 1 until N).foreach { j =>
+        val entry = results(i)(j)
+        // Expected value of entry is 5000
+        assert(entry > 4000 && entry < 6000)
```
Comment (patrick-schultz) [inline hail/src/test/scala/is/hail/expr/ir/EmitStreamSuite.scala:223]: It's safe to tighten this up. I'd suggest 4880 and 5650. The probability that not all entries are within that window is < 1e-5. (Expected value is 5263.)

**Review by patrick-schultz — DISMISSED** [review]

File: hail/src/test/scala/is/hail/expr/ir/EmitStreamSuite.scala (hunk 3 above)
Comment (johnc1231) [reply hail/src/test/scala/is/hail/expr/ir/EmitStreamSuite.scala]: Done, thanks. I mistakenly thought it was 5000.

**Review by patrick-schultz — APPROVED** [review]

**Pull request merged**